File: clients/go/consensus/flagday.go
package consensus

import (
	"fmt"
	"sort"
)

// FlagDayDeployment implements CANONICAL §23.2 height-activation semantics.
//
// Consensus validity MUST depend only on (height >= activation_height). Any version-bit signaling
// is telemetry-only and MUST NOT gate validity.
type FlagDayDeployment struct {
	Name             string
	ActivationHeight uint64
	// Bit is optional and telemetry-only (0..31).
	Bit *uint8
}

func (d FlagDayDeployment) Validate() error {
	if d.Name == "" {
		return fmt.Errorf("flagday: name required")
	}
	if d.Bit != nil && *d.Bit > 31 {
		return fmt.Errorf("flagday: bit out of range: %d", *d.Bit)
	}
	return nil
}

// ValidateDeploymentBitUniqueness checks a set of FlagDayDeployments for
// telemetry-bit reuse conflicts per CANONICAL §23.2.3. Returns a list of
// human-readable warnings. An empty slice means no conflicts detected.
//
// This is non-consensus: bit collisions do not invalidate blocks. Implementations
// SHOULD log returned warnings at startup.
func ValidateDeploymentBitUniqueness(deployments []FlagDayDeployment) []string {
	// Collect deployments that have a telemetry bit assigned.
	type entry struct {
		name       string
		bit        uint8
		reserveEnd uint64 // activation_height + FALLOW_PERIOD
	}
	var withBit []entry
	for _, d := range deployments {
		if d.Bit == nil {
			continue
		}
		end := d.ActivationHeight + FALLOW_PERIOD
		// Handle uint64 overflow (unlikely but defensive).
		if end < d.ActivationHeight {
			end = ^uint64(0)
		}
		withBit = append(withBit, entry{
			name:       d.Name,
			bit:        *d.Bit,
			reserveEnd: end,
		})
	}

	// Sort by bit, then by reserveEnd for deterministic output.
	sort.Slice(withBit, func(i, j int) bool {
		if withBit[i].bit != withBit[j].bit {
			return withBit[i].bit < withBit[j].bit
		}
		return withBit[i].reserveEnd < withBit[j].reserveEnd
	})

	var warnings []string
	for i := 0; i < len(withBit); i++ {
		for j := i + 1; j < len(withBit); j++ {
			if withBit[i].bit != withBit[j].bit {
				break // sorted by bit — no more matches
			}
			// Same bit: check if reservation windows overlap.
			// Deployment j's bit is assigned from height 0 (or whenever it was declared)
			// through deployment j's reserveEnd. Since we only track reserveEnd (upper bound),
			// any two deployments on the same bit whose reserveEnd windows coexist = overlap.
			warnings = append(warnings, fmt.Sprintf(
				"flagday: bit %d reuse overlap between %q (reserved until height %d) and %q (reserved until height %d) — §23.2.3 FALLOW_PERIOD violation",
				withBit[i].bit,
				withBit[i].name, withBit[i].reserveEnd,
				withBit[j].name, withBit[j].reserveEnd,
			))
		}
	}
	return warnings
}

func FlagDayActiveAtHeight(d FlagDayDeployment, height uint64) (bool, error) {
	if err := d.Validate(); err != nil {
		return false, err
	}
	return height >= d.ActivationHeight, nil
}

File: clients/rust/crates/rubin-consensus/src/flagday.rs
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FlagDayDeployment {
    pub name: String,
    pub activation_height: u64,
    // Telemetry-only (0..31).
    pub bit: Option<u8>,
}

impl FlagDayDeployment {
    pub fn validate(&self) -> Result<(), String> {
        if self.name.is_empty() {
            return Err("flagday: name required".to_string());
        }
        if let Some(b) = self.bit {
            if b > 31 {
                return Err(format!("flagday: bit out of range: {b}"));
            }
        }
        Ok(())
    }
}

/// Check a set of FlagDayDeployments for telemetry-bit reuse conflicts per
/// CANONICAL §23.2.3. Returns a list of human-readable warnings. An empty Vec
/// means no conflicts detected.
///
/// This is non-consensus: bit collisions do not invalidate blocks.
/// Implementations SHOULD log returned warnings at startup.
pub fn validate_deployment_bit_uniqueness(deployments: &[FlagDayDeployment]) -> Vec<String> {
    use crate::constants::FALLOW_PERIOD;

    struct Entry {
        name: String,
        bit: u8,
        reserve_end: u64,
    }

    let mut with_bit: Vec<Entry> = deployments
        .iter()
        .filter_map(|d| {
            d.bit.map(|b| Entry {
                name: d.name.clone(),
                bit: b,
                reserve_end: d.activation_height.saturating_add(FALLOW_PERIOD),
            })
        })
        .collect();

    // Sort by bit, then by reserve_end for deterministic output.
    with_bit.sort_by(|a, b| a.bit.cmp(&b.bit).then(a.reserve_end.cmp(&b.reserve_end)));

    let mut warnings = Vec::new();
    for i in 0..with_bit.len() {
        for j in (i + 1)..with_bit.len() {
            if with_bit[i].bit != with_bit[j].bit {
                break; // sorted by bit — no more matches
            }
            warnings.push(format!(
                "flagday: bit {} reuse overlap between {:?} (reserved until height {}) and {:?} (reserved until height {}) — §23.2.3 FALLOW_PERIOD violation",
                with_bit[i].bit,
                with_bit[i].name, with_bit[i].reserve_end,
                with_bit[j].name, with_bit[j].reserve_end,
            ));
        }
    }
    warnings
}

pub fn flagday_active_at_height(d: &FlagDayDeployment, height: u64) -> Result<bool, String> {
    d.validate()?;
    Ok(height >= d.activation_height)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flagday_active_at_height_basic() {
        let d = FlagDayDeployment {
            name: "X".to_string(),
            activation_height: 100,
            bit: Some(5),
        };
        assert!(!flagday_active_at_height(&d, 99).unwrap());
        assert!(flagday_active_at_height(&d, 100).unwrap());
    }

    #[test]
    fn flagday_bit_range() {
        let d = FlagDayDeployment {
            name: "X".to_string(),
            activation_height: 0,
            bit: Some(32),
        };
        let err = flagday_active_at_height(&d, 0).unwrap_err();
        assert_eq!(err, "flagday: bit out of range: 32");
    }

    #[test]
    fn bit_uniqueness_no_conflict() {
        let ds = vec![
            FlagDayDeployment { name: "A".into(), activation_height: 1000, bit: Some(3) },
            FlagDayDeployment { name: "B".into(), activation_height: 5000, bit: Some(5) },
        ];
        assert!(validate_deployment_bit_uniqueness(&ds).is_empty());
    }

    #[test]
    fn bit_uniqueness_same_bit_overlap() {
        let ds = vec![
            FlagDayDeployment { name: "A".into(), activation_height: 1000, bit: Some(3) },
            FlagDayDeployment { name: "B".into(), activation_height: 2000, bit: Some(3) },
        ];
        let w = validate_deployment_bit_uniqueness(&ds);
        assert_eq!(w.len(), 1);
        assert!(w[0].contains("bit 3 reuse overlap"));
    }

    #[test]
    fn bit_uniqueness_no_bit_skipped() {
        let ds = vec![
            FlagDayDeployment { name: "A".into(), activation_height: 1000, bit: None },
            FlagDayDeployment { name: "B".into(), activation_height: 1000, bit: None },
        ];
        assert!(validate_deployment_bit_uniqueness(&ds).is_empty());
    }

    #[test]
    fn bit_uniqueness_three_way_overlap() {
        let ds = vec![
            FlagDayDeployment { name: "A".into(), activation_height: 1000, bit: Some(7) },
            FlagDayDeployment { name: "B".into(), activation_height: 1500, bit: Some(7) },
            FlagDayDeployment { name: "C".into(), activation_height: 2000, bit: Some(7) },
        ];
        assert_eq!(validate_deployment_bit_uniqueness(&ds).len(), 3);
    }

    #[test]
    fn bit_uniqueness_empty() {
        assert!(validate_deployment_bit_uniqueness(&[]).is_empty());
    }
}
